### PR TITLE
Prometheus exporters

### PR DESCRIPTION
**Pull request by MostAwesomeDude**
###### Motivation for this change

I started using Prometheus recently and I needed to export some stuff.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This all appears to work; https://graphs.matador.cloud/dashboard/db/web-status shows off the nginx and Varnish support, while https://graphs.matador.cloud/dashboard/db/storage-overview uses JSON scraping.

### CODE REVIEW

File: pkgs/servers/monitoring/prometheus/varnish-exporter.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# This file was generated by go2nix.
+{ stdenv, buildGoPackage, fetchgit, lib }:
+
+buildGoPackage rec {
+  name = "prometheus_varnish_exporter-${version}";
+  version = "20161126-${stdenv.lib.strings.substring 0 7 rev}";
+  rev = "218cffa79f9b56de6a21a9244d74bdc2ec089473";
```
Comment (Mic92) [inline pkgs/servers/monitoring/prometheus/varnish-exporter.nix]: Is there a reason not to use a released version here? https://github.com/jonnenauha/prometheus_varnish_exporter/releases

File: pkgs/servers/monitoring/prometheus/varnish-exporter.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+  goPackagePath = "github.com/jonnenauha/prometheus_varnish_exporter";
+
+  src = fetchgit {
```
Comment (Mic92) [inline pkgs/servers/monitoring/prometheus/varnish-exporter.nix]: fetchFromGitHub is preferred here.

File: pkgs/servers/monitoring/prometheus/varnish-exporter.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    url = "https://github.com/jonnenauha/prometheus_varnish_exporter";
+    sha256 = "1g78bfrxf6nkp4jd6sgc597xxyvi3z6ibbdn6dpkyzy13379pm8l";
+  };
+
+  goDeps = ./varnish-exporter_deps.nix;
+
+  meta = {
+    homepage = "https://github.com/jonnenauha/prometheus_varnish_exporter";
+    description = "Varnish exporter for Prometheus";
+    license = lib.licenses.mit;
```
Comment (Mic92) [inline pkgs/servers/monitoring/prometheus/varnish-exporter.nix]: Do you want to maintain this package?

File: nixos/modules/services/monitoring/prometheus/nginx-exporter.nix [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  config = mkIf cfg.enable {
+    systemd.services.prometheus-nginx-exporter = {
+      description = "Prometheus exporter for nginx metrics";
+      unitConfig.Documentation = "https://github.com/discordianfish/nginx_exporter";
+      wantedBy = [ "multi-user.target" ];
+      script = ''
+        exec ${pkgs.prometheus-nginx-exporter}/bin/nginx_exporter \
+          -nginx.scrape_uri '${cfg.scrapeURI}' \
+          -telemetry.address :${toString cfg.port} \
+          ${concatStringsSep " \\\n  " cfg.extraFlags}
```
Comment (Mic92) [inline nixos/modules/services/monitoring/prometheus/nginx-exporter.nix]: script is unnecessary here. Please use `ExecStart` directly.

File: nixos/modules/services/monitoring/prometheus/varnish-exporter.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  config = mkIf cfg.enable {
+    systemd.services.prometheus-varnish-exporter = {
+      description = "Prometheus exporter for Varnish metrics";
+      unitConfig.Documentation = "https://github.com/jonnenauha/prometheus_varnish_exporter";
+      wantedBy = [ "multi-user.target" ];
+      script = ''
+        export PATH=${pkgs.varnish}/bin
+        exec ${pkgs.prometheus-varnish-exporter}/bin/prometheus_varnish_exporter \
+          -web.listen-address :${toString cfg.port} \
+          ${concatStringsSep " \\\n  " cfg.extraFlags}
```
Comment (Mic92) [inline nixos/modules/services/monitoring/prometheus/varnish-exporter.nix]: Same here. The `PATH` can be extended using the `path` property.

File: pkgs/servers/monitoring/prometheus/json-exporter.nix [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# This file was generated by go2nix.
+{ stdenv, buildGoPackage, fetchgit, lib }:
+
+buildGoPackage rec {
+  name = "prometheus-json-exporter-${version}";
+  version = "20160913-${stdenv.lib.strings.substring 0 7 rev}";
```
Comment (Mic92) [inline pkgs/servers/monitoring/prometheus/json-exporter.nix]: Please use the following version scheme for packages with releases: `prometheus-json-exporter-unstable-2016-09-13`

File: nixos/modules/services/monitoring/prometheus/json-exporter.nix [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  config = mkIf cfg.enable {
+    systemd.services.prometheus-json-exporter = {
+      description = "Prometheus exporter for JSON over HTTP";
+      unitConfig.Documentation = "https://github.com/kawamuray/prometheus-json-exporter";
+      wantedBy = [ "multi-user.target" ];
+      script = ''
+        exec ${pkgs.prometheus-json-exporter}/bin/prometheus-json-exporter \
+          --port ${toString cfg.port} \
+          ${cfg.url} ${cfg.configFile} \
+          ${concatStringsSep " \\\n  " cfg.extraFlags}
```
Comment (Mic92) [inline nixos/modules/services/monitoring/prometheus/json-exporter.nix]: same here: `script` -> `ExecStart`

File: pkgs/servers/monitoring/prometheus/json-exporter.nix [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  rev = "d45e5ebdb08cb734ad7a8683966032af1d91a76c";
+
+  goPackagePath = "github.com/kawamuray/prometheus-json-exporter";
+
+  src = fetchgit {
```
Comment (Mic92) [inline pkgs/servers/monitoring/prometheus/json-exporter.nix]: `fetchFromGitHub`

**Review by Mic92 — CHANGES_REQUESTED** [review]

**Review by fpletz — APPROVED** [review]
Thanks! Great work! :+1:

But could you please squash your fixes into the previous commits?

**Comment by globin** [comment]
Cleaned this up a bit and added stuff we had lying around in our fork.

**Pull request merged**